### ISSUE TITLE
FetchBy is gone, advice is needed

### ISSUE THREAD

**Issue by konstk** [schotime/NPoco#170]
I have tried to update to the latest version, but with the FetchBy gone we cannot use it. The Query<T> simply does not have the same power, a lot of our code is now broken.

We were able to construct Linq queries to pass the by to ANY IDatabase (via repositories), where we could optionally make them part of a transaction or execute against the replica. We could control the execution of the query or check the validity. Now, with QueryProvider<T> directly coupled to specific IDatabase right from the start and directly exposing ToList() etc methods, all DAL logic cannot be contained within repositories. Anyone who makes use of IQueryProvider<T> can just call ToList() whenever and wherever they feel like and it will hit the database (possibly wrong replica).
```C#
Func<SqlExpression<Class1>, SqlExpression<Class1>> sql = m => m.Where(x => x.Nr == nr);
if (!includeExpired)
    sql += m => m.Where(x => x.EndDate >= DateTime.Today || x.EndDate == null);
...
...
Database1.FetchBy(sql);
Database2.FetchBy(sql);
```
How do we do it with suggested Query<T> ? Examples?

**Comment by konstk** [comment]
Oh, and writing unit tests for Query<T> / IQueryProvider<T> requires far more code, mock setups, etc. I have to write setups for all these Where OrderBy etc etc etc whatever is use in the code and make a mock with setups from IDatabase too...

**Comment by schotime** [comment]
Its fairly easy to compose queries.

```
Expression<Func<Type, bool>> sql = m => m.Where(x => x.Nr == nr);
if (!includeExpired)
    sql = sql.Where(x => x.EndDate >= DateTime.Today || x.EndDate == null);

Database1.Query<Type>(sql).ToList();
Database2.Query<Type>(sql).ToList();
```

**Comment by konstk** [comment]
Adam

Thank you very much for your answer. However, for some reason, it doesn't work for me.

Expression<Func<Type, bool>> sql = m => m.Where(x => x.Nr == nr);

There is no Where in "m." part. After "m." I get properties from the Type.

Also, the following Query<T>(sql) function doesn't have an overload that takes Expression as a parameter:

Database.Query<Type>(sql).ToList()

I feel like I am missing something obvious. The Expression in your example is the standard Expression object from System.Linq.Expressions namespace?

**Comment by schotime** [comment]
My bad. Firstly Type is your type eg. User
Then you can use this.

```csharp
var where = NPoco.Expressions.PredicateBuilder.True<User>();
where = where.And(x => x.Name == "Name");
var query = db.Query<User>().Where(where).ToList();
```

**Comment by konstk** [comment]
Adam

Thank you for the clarification! However, with ```FetchBy``` we were able to construct complete query with ```Where```, ```OrderBy```, ```ThenBy```, ```Limit```, etc. completely independently. 

For example:

```csharp
FetchBy(
    m =>  m.Where(x => x.Id == -1 || x.Id == Id)
             .Where(x => x.Code == code)
             .OrderBy(x => x.Birthday)
             .ThenByDescending(x => x.Gender)
             .Limit(20)
    ).ToList();
```

**Comment by schotime** [comment]
I am currently in the middle of the 3.0 release which will have significant changes so I will try and include it for you. Best I can do for now.

**Comment by konstk** [comment]
Adam

Thank you! I hope it'll work! 

We use NPoco quite heavily in our systems, and re-writing 70% of DAL and Unit Tests is not a small task. What is inconvenient for a small companies, can become impractical for large enterprises (too much critical code will be affected).

**Comment by schotime** [comment]
There you go ;)
https://github.com/schotime/NPoco/commit/1c5e56996677e28caaef4d64eec958df0a332183#diff-296728837ef3bc31f963f5bd85d46353

Its working! Going to release an alpha soon, but need to do more regression testing. Its looking good though.

**Comment by konstk** [comment]
Looks good! I tried this version (downloaded from GitHub), but I get 
```
An unhandled exception of type 'System.Security.VerificationException' occurred in NPoco.dll
Additional information: Operation could destabilize the runtime.
```
Indeed, some testing is needed :-)

I really hope you will make conditional code specifically for .NET 4.5, so no external packages must be dragged along NPoco... for async/await, etc. Three Nuget packages just to target .NET 4.0. Also, strange reference to Microsoft.JScript in the NPoco project.

**Comment by konstk** [comment]
This exception occurs in:
```csharp
/// <summary>
/// Sets the property for the specified target.
/// </summary>
/// <param name="target">Target object.</param>
/// <param name="value">Value to set.</param>
public void Set(object target, object value)
{
    SetDelegate(target, value);
}
```

**Comment by schotime** [comment]
No its not that bad. 
All tests are passing and if you use the build script you will get the different versions compiled properly.

What were you trying to do?

**Comment by konstk** [comment]
Sorry, I wasn't clear about .NET 4.5. I just wish you will have a separate Nuget package with NPoco dll without extra dependencies.

As for the exception, I didn't do anything special:
```csharp
var gem2 = database.Fetch<MyClass>().First();
```
May be something is wrong with my setup though. Any attempt to retrieve data leads to ```Operation could destabilize the runtime.``` exception in my test project.

**Comment by schotime** [comment]
There are no extra dependencies for .net4.5.
What does MyClass look like?

**Comment by konstk** [comment]
MyClass is as simple as it gets:

```csharp
[TableName("TestClasses")]
[PrimaryKey("ID")]
[DebuggerDisplay("MyClass [Id={Id}, Name={Name}]")]
[ExplicitColumns]
[ExcludeFromCodeCoverage]
[GeneratedCode("T4 for NPoco by Konstantin", "1.9.1")]
public partial class MyClass
{
	[Column("ID")]
	public virtual int Id { get; set; }
	[Column("Name")]
	public virtual string Name { get; set; }
}
```

**Comment by schotime** [comment]
Can you run the tests?
Are you running the latest commit of dev3.0?
I am trying to get to the bottom of this but I need some more information.

**Comment by konstk** [comment]
I just downloaded a fresh copy of dev3.0.

12 tests failed for various reasons, but 3 out of 12 failed with the same exception:

Test Name:	GenerateCloneWIthPrivateConstructor
Test FullName:	NPoco.Tests.SnapshotterTests.GenerateCloneWIthPrivateConstructor
Test Source:	c:\Temp\NPoco-dev3.0\src\NPoco.Tests\SnapshotterTests.cs : line 99
Test Outcome:	Failed
Test Duration:	0:00:00,016

Result Message:	System.Security.VerificationException : Operation could destabilize the runtime.
Result StackTrace:	
at DoClone(UserWithPrivateParamLessConstructor )
at NPoco.Snapshot`1.Clone(T myObject, PocoData data) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 74
at NPoco.Snapshot`1..ctor(PocoData data, T original) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 37
at NPoco.Snapshotter.StartSnapshot[T](IDatabase d, T obj) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 17
at NPoco.Tests.SnapshotterTests.GenerateCloneWIthPrivateConstructor() in c:\Temp\NPoco-dev3.0\src\NPoco.Tests\SnapshotterTests.cs:line 101

Test Name:	GetDiffIncludingLists
Test FullName:	NPoco.Tests.SnapshotterTests.GetDiffIncludingLists
Test Source:	c:\Temp\NPoco-dev3.0\src\NPoco.Tests\SnapshotterTests.cs : line 133
Test Outcome:	Failed
Test Duration:	0:00:00,002

Result Message:	System.Security.VerificationException : Operation could destabilize the runtime.
Result StackTrace:	
at DoClone(SnapshotOnClass )
at NPoco.Snapshot`1.Clone(T myObject, PocoData data) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 74
at NPoco.Snapshot`1..ctor(PocoData data, T original) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 37
at NPoco.Snapshotter.StartSnapshot[T](IDatabase d, T obj) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 17
at NPoco.Tests.SnapshotterTests.GetDiffIncludingLists() in c:\Temp\NPoco-dev3.0\src\NPoco.Tests\SnapshotterTests.cs:line 138

Test Name:	GetDiffIncludingObjectPropertiesOnlyByReferenceThough
Test FullName:	NPoco.Tests.SnapshotterTests.GetDiffIncludingObjectPropertiesOnlyByReferenceThough
Test Source:	c:\Temp\NPoco-dev3.0\src\NPoco.Tests\SnapshotterTests.cs : line 112
Test Outcome:	Failed
Test Duration:	0:00:00,001

Result Message:	System.Security.VerificationException : Operation could destabilize the runtime.
Result StackTrace:	
at DoClone(SnapshotOnClass )
at NPoco.Snapshot`1.Clone(T myObject, PocoData data) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 74
at NPoco.Snapshot`1..ctor(PocoData data, T original) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 37
at NPoco.Snapshotter.StartSnapshot[T](IDatabase d, T obj) in c:\Temp\NPoco-dev3.0\src\NPoco\Snapshotter.cs:line 17
at NPoco.Tests.SnapshotterTests.GetDiffIncludingObjectPropertiesOnlyByReferenceThough() in c:\Temp\NPoco-dev3.0\src\NPoco.Tests\SnapshotterTests.cs:line 120

**Comment by konstk** [comment]
I must admit this exception we had with NPoco before, when we tried to create Stapshot<T> from a proxy class (we proxy pocos with Castle Proxy to add some functionality on the fly). But this test I run directly from your code (and solution file). My small test also does NOT use proxies, just pocos out of the box.

**Comment by schotime** [comment]
I don't care that the snapshots fail at the moment. That is due to supporting complex object properties.

The point is you were able to run the majority of the sql tests, so there must be something different between what you are doing and what the test is doing.

**Comment by konstk** [comment]
Interesting. I removed "virtual" from properties and it works now.

**Comment by schotime** [comment]
Ok. cool. Well that gives me something to work with.

**Comment by schotime** [comment]
Think I have fixed it. Can you fetch again.

**Comment by konstk** [comment]
This seems to work. But I noticed a strange side effect (may be it happend before).

I have another class, for simplicity let's call it ```House```:

```csharp
public partial class House
{
    [Column("ID")]
    public virtual int Id { get; set; }
    [Column("House")]
    public virtual string Name { get; set; }
}
````

Note: the mapped column of the property Name has the same name as House class. When I use
```database.Fetch<House>().First();```
The name property has Null value.

It works, however, if I get a ```List<House>``` from ```Query<House>().From(query).ToList()```. Name property of all classes are filled in.

I tested it on several classes and whenever Column attribute has the same value as class name, this and only this property is NOT filled in.
